### ISSUE TITLE
[feature] Expose event to listener functions

### ISSUE THREAD

**Issue by bcallaghan-fri** [sibnerian/electron-promise-ipc#8]
When using the built-in IPC methods, the first parameter to the listener function is the event. That event object is currently hidden when using `electron-promise-ipc`. It would be useful if that object were passed into the callbacks. That would allow much greater flexibility for developers to write the same logic that they could otherwise write with vanilla Electron. I am not looking to alter the request/response logic or send my own responses. There are a few properties found only on `event.sender` that are useful to me.

**Comment by sibnerian** [comment]
Can you tell me a little more about your use-case? I wrote this library as
an abstraction over the event-based IPC flow.  "Vanilla" electron still
works fine if you want to use events.

**Comment by bcallaghan-fri** [comment]
I'm currently writing an Electron application that makes heavy use of promises in both the main and renderer processes. Nearly all of the IPC "events" in this app are actually single-fire RPC functions. I found myself using the following pattern extensively.
```js
return new Promise((resolve, reject) => {
    ipcRenderer.once('response', (event, data) => resolve(data));
    ipcRenderer.send('request');
});
```
This library provides a very clean abstraction over this pattern, as well as more robust error handling and re-entrancy. My existing functions have only used the `event` object to send responses. I now have a similar function that opens a second BrowserWindow, waits for user interaction, and returns a single value back to the original renderer process. This second window should be a child modal of the first, but that requires a reference to the original window. I can get this reference through `event.sender.getOwnerBrowserWindow()` in the standard `ipcMain` callback. I would like to keep the Promise-based abstraction while also gaining the ability to use the `event.sender` object.

**Comment by sibnerian** [comment]
OK, I think I see the issue:

1. ipcRenderer makes a call to ipcMain requesting data.
2. ipcMain handles this call with .on() and wants to open a pop up window
to get that data (this is the part I’m not sure about)
3. To open that window it needs a handle to event.sender, which is not
exposed in the current API.

Is that correct?

I think we can figure out an API that works for this. Would it be as simple
as passing the event as the last argument to the .on() listener function?
So instead of calling `listener(...dataArgs)` it’d call
`listener(...dataArgs, event)`.

**Comment by sibnerian** [comment]
OK, I think I see the issue:

1. ipcRenderer makes a call to ipcMain requesting data.
2. ipcMain handles this call with .on() and wants to open a pop up window
to get that data (this is the part I’m not sure about)
3. To open that window it needs a handle to event.sender, which is not
exposed in the current API.

Is that correct?

I think we can figure out an API that works for this. Would it be as simple
as passing the event as the last argument to the .on() listener function?
So instead of calling `listener(...dataArgs)` it’d call
`listener(...dataArgs, event)`.

**Comment by sibnerian** [comment]
@bcallaghan-fri friendly ping!

**Comment by bcallaghan-fri** [comment]
Sorry for the delayed response. I usually only work on this project on the weekends.

Yes, you have the process correct. Passing `event` as the final argument should work. The main process handler would then look something like this.
```js
promiseIpc.on('my-event', (arg1, arg2, event) => {
    const window = BrowserWindow.fromWebContents(event.sender);
    // Additional logic using the sender's window
});
```

**Comment by sibnerian** [comment]
Resolved in #9.
